core-bundle/src/Menu/FrontendMenuBuilder.php

```diff
@@ -0,0 +1,362 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Menu;
+
+use Contao\CoreBundle\Event\FrontendMenuEvent;
+use Contao\CoreBundle\Framework\Adapter;
+use Contao\CoreBundle\Monolog\ContaoContext;
+use Contao\CoreBundle\Routing\Page\PageRegistry;
+use Contao\CoreBundle\Security\Authentication\Token\TokenChecker;
+use Contao\CoreBundle\Security\ContaoCorePermissions;
+use Contao\Database;
+use Contao\Date;
+use Contao\PageModel;
+use Contao\StringUtil;
+use Doctrine\DBAL\Connection;
+use Knp\Menu\FactoryInterface;
+use Knp\Menu\ItemInterface;
+use Psr\Log\LoggerInterface;
+use Psr\Log\LogLevel;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Routing\Exception\ExceptionInterface;
+use Symfony\Component\Security\Core\Security;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+
+class FrontendMenuBuilder
+{
+    private FactoryInterface $factory;
+    private RequestStack $requestStack;
+    private EventDispatcherInterface $dispatcher;
+    private Connection $connection;
+    private PageRegistry $pageRegistry;
+    /**
+     * @var Adapter<PageModel>
+     */
+    private Adapter $pageModelAdapter;
+    private TokenChecker $tokenChecker;
+    private Security $security;
+    private LoggerInterface $logger;
+    private Database $database;
+
+    /**
+     * @param Adapter<PageModel> $pageModelAdapter
+     */
+    public function __construct(FactoryInterface $factory, RequestStack $requestStack, EventDispatcherInterface $dispatcher, Connection $connection, PageRegistry $pageRegistry, Adapter $pageModelAdapter, TokenChecker $tokenChecker, Security $security, LoggerInterface $logger, Database $database)
+    {
+        $this->factory = $factory;
+        $this->requestStack = $requestStack;
+        $this->dispatcher = $dispatcher;
+        $this->connection = $connection;
+        $this->pageRegistry = $pageRegistry;
+        $this->pageModelAdapter = $pageModelAdapter;
+        $this->tokenChecker = $tokenChecker;
+        $this->security = $security;
+        $this->logger = $logger;
+        $this->database = $database;
+    }
+
+    public function getMenu(ItemInterface $root, int $pid, array $options = []): ItemInterface
+    {
+        $options = array_replace([
+            'showHidden' => false,
+            'showProtected' => false,
+            'showLevel' => 0,
+            'hardLimit' => false,
+            'isSitemap' => false,
+        ], $options);
+
+        $pages = $this->getPages($pid, $options);
+
+        $request = $this->requestStack->getCurrentRequest();
+        /** @var PageModel|null $currentPage */
+        $currentPage = $request->attributes->get('pageModel');
```

Comment:
In an ESI request this attribute will only contain the page ID. Until we have a general service to retrieve the current page model, I think we still need to use `$GLOBALS['objPage']` here. /cc @aschempp

Reply:
or use the PageModelAdapter to fetch the page

Reply:
awaiting feedback from Andy concerning 02f81fd5b7657ee522b428516f5e0b7a5b16ec98.

Reply:
May be should put this in a common service. Currently we have `ContaoContext::getPageModel` and a `AbstractFragmentController::getPageModel` which do (almost) exactly the same thing, i.e. we have duplicated code.

Reply:
🤔 yes, someone decide pls

Reply:
I had the same issue in the image studio code as well and ended up using `$GLOBALS['objPage']` because there was no better solution. Big :+1: for a service from my side.

Reply:
Replicated the code in f5105224bb51236ad27408f64a21239f32cd11dd again. But I am not copying the tests for this method 🤣 We need a general service the other day.

Reply:
see my suggestion in https://github.com/contao/contao/pull/3930